Refactored to use implicit insert and remove insert_type method from context object  #3395

mostafa630 opened the pull request:
This solution resolves issue #3395 by refactoring the code to use implicit_insert_type instead of insert_type, improving clarity and eliminating the insert_type method from the context object.
Changed files:
-typecheck/rust-substitution-mapper.cc: Updated to use insert_implicit_type instead of insert_type.
-typecheck/rust-hir-type-check.h: Removed the insert_type declaration from the TypeCheckContext class.
-typecheck/rust-typecheck-context.cc: Removed the insert_type definition.

powerboat9 commented:
You don't have to keep recreating the PR, you can just force-push

powerboat9 commented:
Looks good, just missing the removal of a few `insert_type` calls

mostafa630 commented:
> Looks good, just missing the removal of a few `insert_type` calls

Should I only remove the remaining ones, or should I replace them with insert_implicit_type?

powerboat9 commented:
Replace them with `insert_implicit_type`

dkm commented:
FYI, changelog is not really correct. You should have things like:
```
   * typecheck/rust-substitution-mapper.cc (SubstMapperInternal::Resolve): use insert_implicit_type.
   * typecheck/rust-typecheck-context.cc (TypeCheckContext::insert_type): Remove.
   (TypeCheckContext::compute_inference_variables): use insert_implicit_type.
```
...
You're not supposed to only point at the file where you've made changes, but to more precise "item" (function, type definition, etc).
Also, instead of repeating the same thing in sequence, you can use the idiomatic "Likewise" (have a look at existing entries in `git log`). Do not hesitate to ask for help if needed :)

mostafa630 commented:
@dkm can I use clang-format-13 or Imust use version 10  because i has some problems in installing version 10

mostafa630 commented:
> I use version 19, which only very rarely disagrees with version 10. You can always try it with version 13 and see if the CI complains

@powerboat9 Thanks! I tried it with version 13, and it works well. All tests passed successfully.

P-E-P commented:
@dkm I wonder if pseudonyms are allowed for sign off

philberty commented on gcc/rust/typecheck/rust-typecheck-context.cc:
```diff
@@ -79,17 +79,6 @@ TypeCheckContext::get_builtins () const
-  rust_assert (type != nullptr);
-  NodeId ref = mappings.get_nodeid ();
-  HirId id = mappings.get_hirid ();
-  node_id_refs[ref] = id;
```
can you check to see if we can remove the node_id_refs map? Is it still used elsewhere?

mostafa630 replied:
The following functions utilize the node_id_refs map:

-`TypeCheckContext::lookup_builtin(NodeId id, TyTy::BaseType **type)`
-`TypeCheckContext::insert_builtin(HirId id, NodeId ref, TyTy::BaseType *type)`
-`TypeCheckContext::insert_type_by_node_id(NodeId ref, HirId id)`
-`TypeCheckContext::lookup_type_by_node_id(NodeId ref, HirId *id)`

but , these functions are only exist in two files:
-rust-hir-type-check.h
-rust-typecheck-context.cc

Additionally, while TypeCheckContext::lookup_builtin is used elsewhere, the version being used is not the one mentioned above. Instead, the used version is an overloaded function:
-`TypeCheckContext::lookup_builtin (std::string name, TyTy::BaseType **type)`

Given this, we can safely remove the `node_id_refs` map and all functions that depend on it, as doing so will not impact the code , unless the project requires them in the future or in a context that is not clear to me.

mostafa630 replied:
sorry after more research i found that there are files in other  folders like  resolve call some function that use 
`node_id_refs`  map
and that is example : 
in the file  : rust-name-resolver.cc  in the resolve folder  it call that  : 
```
tyctx->insert_builtin (unit_tyty->get_ref (), unit_type->get_node_id (),
			 unit_tyty);
```

and that `insert_builtin` uses node_id_refs map

philberty replied:
no problem we can probably remove that i will make a new issue then

philberty reviewed (changes requested):
LGTM can you just verify if we still need that node_id_refs map?

philberty reviewed (changes requested):
I think you need to fix your sign-off for DCO you are using a pseudonym 

https://gcc.gnu.org/dco.html

mostafa630 commented:
> I think you need to fix your sign-off for DCO you are using a pseudonym
> 
> https://gcc.gnu.org/dco.html

I changed the sign-off to match the DCO.

dkm commented:
> > I think you need to fix your sign-off for DCO you are using a pseudonym
> > https://gcc.gnu.org/dco.html
> 
> I changed the sign-off to match the DCO.

erm, I thought they were going to relax this rule. At least, it was mentioned during last Cauldron I think.

(please note that my initial comment on changelog section still holds...)

mostafa630 commented:
> > > I think you need to fix your sign-off for DCO you are using a pseudonym
> > > https://gcc.gnu.org/dco.html
> > 
> > 
> > I changed the sign-off to match the DCO.
> 
> erm, I thought they were going to relax this rule. At least, it was mentioned during last Cauldron I think.
> 
> (please note that my initial comment on changelog section still holds...)

sorry , but do you mean that i should  make this  : 
	* typecheck/rust-hir-type-check.h: Remove declaration of insert_type method.
like that : 
      * typecheck/rust-typecheck-context.cc
	  (TypeCheckContext::insert_type): Remove.
	  
or do you mean that i don't mentioned which  function exactly i changed it : 
	* typecheck/rust-tyty-subst.cc: Replace all insert_type with insert_implicit_type.

(i don't mentioned functions because there are many of them so the change log will be huge)

philberty commented:
Your git signoff line needs to mention your full name it cant be a username

mostafa630 commented:
> Your git signoff line needs to mention your full name it cant be a username

Okay, I will use that: Mostafa Mahmoud Younis.
But is there any further update in the message change log?
